Library/Homebrew/cli/args.rb

```diff
@@ -12,6 +12,135 @@ def initialize(argv:)
         super
         @argv = argv
       end
+
+      def to_cli_option(name)
+        if name.length == 2
+          "-#{name.tr("?", "")}"
+        else
+          "--#{name.tr("_", "-").tr("?", "")}"
+        end
+      end
+
+      def options_only
+        to_h.keys
+            .map(&:to_s)
+            .reject { |name| %w[argv remaining].include?(name) }
+            .map(&method(:to_cli_option))
+            .select { |arg| arg.start_with?("-") }
+      end
+
+      def flags_only
+        to_h.keys
+            .map(&:to_s)
+            .reject { |name| %w[argv remaining].include?(name) }
+            .map(&method(:to_cli_option))
+            .select { |arg| arg.start_with?("--") }
+      end
+
+      def named
+        remaining
+      end
+
+      def formulae
+        require "formula"
+        # TODO: use @instance variable to ||= cache when moving to CLI::Parser
+        (downcased_unique_named - casks).map do |name|
+          if name.include?("/") || File.exist?(name)
+            Formulary.factory(name, spec)
+          else
+            Formulary.find_with_priority(name, spec)
+          end
+        end.uniq(&:name)
+      end
+
+      def resolved_formulae
+        require "formula"
+        # TODO: use @instance variable to ||= cache when moving to CLI::Parser
+        (downcased_unique_named - casks).map do |name|
```

Comment:
`@resolved_formulae ||= (downcased_unique_named - casks).map do |name|` and remove TODO

Reply:
I've already tried exactly this but it was breaking a lot of tests. Let me try again.

Reply:
You may need to add a `Homebrew.args.reset` or similar which will reset the instance variables and have it in https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/spec_helper.rb#L135-L208

In general there's some ARGV stuff there that may need similar code for Homebrew.args

Reply:
I think we've hit a cyclic dependency here. 
[formula_options](https://github.com/Homebrew/brew/pull/6433/files#diff-e0e03670c451d934ae1d79adfff63cacL160) method is used to allow extra arguments for formulae see - [deps.rb#L55](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/deps.rb#L55) and [install.rb#L90](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/install.rb#L90) but formula options itself needs to know what is passed on command line to define what options can be allowed on command line!

Reply:
In that case you might want to allow e.g. `ARGV` to be passed in in some circumstances for initialisation so it's not operating on itself? Would that work?

Reply:
Yeah, if we limit the usage of `ARGV` to this one method we could make this work.

Reply:
I would suggest it's passed in as an argument (a default one if necessary) to make testing easier and just treated as an Array. The issue we're trying to address isn't so much the use of `ARGV` but removing our patches on `ARGV`.